cranelift-codegen/src/regalloc/reload.rs

```diff
@@ -402,11 +404,30 @@ impl<'a> Context<'a> {
         if let Some(constraints) = constraints {
             for (argidx, (op, &arg)) in constraints.ins.iter().zip(args).enumerate() {
                 if op.kind != ConstraintKind::Stack && self.liveness[arg].affinity.is_stack() {
-                    self.candidates.push(ReloadCandidate {
-                        argidx,
-                        value: arg,
-                        regclass: op.regclass,
-                    })
+                    if let Some(sigref) = self.cur.func.dfg.call_signature(inst) {
+                        let sig = &self.cur.func.dfg.signatures[sigref];
+                        if let ArgumentLoc::Reg(reg) = sig.params[argidx].location {
+                            self.candidates.push(ReloadCandidate {
+                                argidx,
+                                value: arg,
+                                reg: (op.regclass, Some(reg)),
+                            });
+                            continue;
+                        }
+                    };
+                    if let ConstraintKind::FixedReg(reg) = op.kind {
+                        self.candidates.push(ReloadCandidate {
+                            argidx,
+                            value: arg,
+                            reg: (op.regclass, Some(reg)),
+                        })
+                    } else {
+                        self.candidates.push(ReloadCandidate {
+                            argidx,
+                            value: arg,
+                            reg: (op.regclass, None),
+                        })
+                    }
```

Comment:
The only thing that differs between the then and else branches of this if is the second element of `reg`'s initializer. Would it make sense to factor out?
```rust
let unit = if let ConstraintKind::FixedReg(reg) = op.kind {
    Some(reg)
} else {
    None
};
self.candidates.push(ReloadCandidate {
    argidx,
    value: arg,
    reg: (op.regclass, unit),
};

Reply:
yes! I did the same with the `call_signature` if above this region as well.